src/cpp/gnode.cpp

```diff
@@ -64,35 +65,47 @@ t_gnode::t_gnode(const t_gnode_recipe& recipe)
     }
 }
 
-t_gnode::t_gnode(const t_schema& portschema)
+t_gnode::t_gnode(const t_gnode_options& options)
     : m_mode(NODE_PROCESSING_SIMPLE_DATAFLOW)
-    , m_tblschema(portschema.drop({"psp_op", "psp_pkey"}))
-    , m_ischemas(t_schemavec{portschema})
+    , m_gnode_type(options.m_gnode_type)
+    , m_tblschema(options.m_port_schema.drop({"psp_op", "psp_pkey"}))
     , m_init(false)
     , m_id(0)
     , m_pool_cleanup([]() {})
 {
     PSP_TRACE_SENTINEL();
     LOG_CONSTRUCTOR("t_gnode");
+
     std::vector<t_dtype> trans_types(m_tblschema.size());
     for (t_uindex idx = 0; idx < trans_types.size(); ++idx)
     {
         trans_types[idx] = DTYPE_UINT8;
     }
 
+    t_schema port_schema(options.m_port_schema);
+    if (m_gnode_type == GNODE_TYPE_IMPLICIT_PKEYED) {
+
+        // Make sure that gnode type is consistent with input schema
+        if (port_schema.has_column("psp_pkey")) {
+            PSP_COMPLAIN_AND_ABORT("gnode type specified as implicit pkey, however input schema has psp_pkey column");
+        }
+        port_schema = t_schema{{"psp_op", "psp_pkey"}, {DTYPE_UINT8, DTYPE_INT64}} + port_schema;
+    }
+
```

Comment:
Add a similar check for pkeyed gnodes to ensure that they have a psp_pkey and psp_op.